File: TankBot/src/Commands/DriveStraight.cpp
#include "DriveStraight.h"

// magic P = 0.9, cycles per period = 6
#define P 0.54
#define I 0.18
#define D 0.405
#define TOL 0.1

static int cycle_count = 0;

DriveStraight::DriveStraight(double d)  : CommandBase("DriveStraight"),
	pid(P,I,D,this,this)
{
	Requires(driveTrain.get());
	distance=d;
	tolerance=TOL;
    std::cout << "new DriveStraight("<<d<<")"<< std::endl;
    frc::SmartDashboard::PutNumber("P",P);
    frc::SmartDashboard::PutNumber("I",I);
    frc::SmartDashboard::PutNumber("D",D);
    frc::SmartDashboard::PutNumber("TOL",TOL);
    frc::SmartDashboard::PutNumber("Target",distance);
}

// Called just before this Command runs the first time
void DriveStraight::Initialize() {
	cycle_count = 0;
	double p = frc::SmartDashboard::GetNumber("P",P);
	double i = frc::SmartDashboard::GetNumber("I",I);
	double d = frc::SmartDashboard::GetNumber("D",D);
	tolerance=frc::SmartDashboard::GetNumber("TOL",tolerance);
	distance = frc::SmartDashboard::GetNumber("Target",distance);
	last_ontarget=false;
	pid.SetPID(p,i,d);
	driveTrain->Reset();
  	pid.Reset();
	pid.SetSetpoint(distance);
	pid.SetAbsoluteTolerance(tolerance);
	pid.Enable();
	//pid.SetToleranceBuffer(5);
	driveTrain->Enable();
	std::cout << "DriveStraight Started .."<< std::endl;
}

// Called repeatedly when this Command is scheduled to run
void DriveStraight::Execute() {
}

// Make this return true when this Command n o longer needs to run execute()
bool DriveStraight::IsFinished() {
	bool new_target=pid.OnTarget();
	if(new_target && last_ontarget){
		return true;
	}
	last_ontarget=new_target;
	return false;
}

// Called once after isFinished returns true
void DriveStraight::End() {
	pid.Disable();
    std::cout << "DriveStraight End" << std::endl;
}

// Called when another command which requires one or more of the same
// subsystems is scheduled to run
void DriveStraight::Interrupted() {
	End();
}

//#define DEBUG_COMMAND
#define PRINT_CYCLES

double DriveStraight::PIDGet() {
	double s=driveTrain->GetDistance();
#ifdef DEBUG_COMMAND
    std::cout << "DriveStraight::PIDGet("<<s<<")"<< std::endl;
#endif
	return s;
}

void DriveStraight::PIDWrite(double d) {
#ifdef DEBUG_COMMAND
    std::cout << "DriveStraight::PIDWrite("<<d<<")"<< std::endl;
#endif
#ifdef PRINT_CYCLES
    if(cycle_count > 0){
    	std::cout << cycle_count << " "<< d << std::endl;
    }
    cycle_count++;
#endif
	driveTrain->TankDrive(d,d);
}

File: 2018FRCbot/src/org/usfirst/frc/team159/robot/commands/DriveStraight.java
package org.usfirst.frc.team159.robot.commands;

import org.usfirst.frc.team159.robot.Robot;

import edu.wpi.first.wpilibj.PIDController;
import edu.wpi.first.wpilibj.PIDOutput;
import edu.wpi.first.wpilibj.PIDSource;
import edu.wpi.first.wpilibj.PIDSourceType;
import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj.command.Command;

/**
 *
 */
public class DriveStraight extends Command implements PIDSource, PIDOutput {
	private PIDController pid;
	private PIDSourceType type = PIDSourceType.kDisplacement;
	private final double distance;
	private static final double tolerance = 0.1;
	private static final boolean lastOntarget = false;
	private boolean started = false;
	private static final boolean debug = true;
	private static final double P = 0.25;
	private static final double I = 0.002;
	private static final double D = 0.0;
	private static final double TOL = 0.05;
	private static final double intime = 0.025;
	private Timer timer = new Timer();

	public DriveStraight(double distance) {
		this.distance = distance;
		pid = new PIDController(P, I, D, this, this);
		pid.setAbsoluteTolerance(TOL);
		requires(Robot.driveTrain);
	}

	// Called just before this Command runs the first time
	protected void initialize() {
		System.out.printf("drivestraight.initialize distance = %f\n", distance);
		pid.reset();
		pid.setSetpoint(distance);
		pid.disable();
		Robot.driveTrain.enable();
		Robot.driveTrain.reset();
		started = false;
		timer.start();

	}

	// Called repeatedly when this Command is scheduled to run
	protected void execute() {
		if (!started && timer.get() > intime && !pid.isEnabled()) {
			pid.reset();
			pid.enable();
			started = true;
		}
	}

	// Make this return true when this Command no longer needs to run execute()
	protected boolean isFinished() {
		if (timer.get() > 3.0) {
			printFinishedMessage();
			return true;
		}
		return pid.onTarget();
	}

	// Called once after isFinished returns true
	protected void end() {
		Robot.driveTrain.disable();
		pid.disable();
		started = false;
		
		printEndMessage();
	}

	// Called when another command which requires one or more of the same
	// subsystems is scheduled to run
	protected void interrupted() {
		end();
	}

	@Override
	public void setPIDSourceType(PIDSourceType pidSource) {
		type = pidSource;
	}

	@Override
	public PIDSourceType getPIDSourceType() {
		return type;
	}

	@Override
	public double pidGet() {
		if (!started) {
			return 0;
		}
		if (debug) {
			double timeval = timer.get();
			System.out.printf("tm=%f l=%g r=%g d=%g\n", timeval * 1000, Robot.driveTrain.getLeftDistance(),
					Robot.driveTrain.getRightDistance(), Robot.driveTrain.getDistance());
		}
		return Robot.driveTrain.getDistance();
	}

	@Override
	public void pidWrite(double output) {
		if (started) {
			Robot.driveTrain.tankDrive(output, output);
		}
	}
	
	private void printEndMessage() {
		System.out.println("drivestraight.end");
	}
	
	private void printFinishedMessage() {
		System.out.println("timer expired");
	}
}
